### PR TITLE
Add support for requesting GPUs

### PR DESCRIPTION
**Pull request by tkanng**
Hi, this PR adds support for requesting nvidia GPUs, according to https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/426. 

It might be useful for ML programs. And I'm not sure it's good enough, so please help me check it again. 

Thanks!

### CODE REVIEW

**Comment by tkanng** [comment]
I signed it!

File: pkg/apis/sparkoperator.k8s.io/v1beta1/types.go [hunk 1]
```diff
@@ -346,6 +346,9 @@ type SparkPodSpec struct {
 	// MemoryOverhead is the amount of off-heap memory to allocate in cluster mode, in MiB unless otherwise specified.
 	// Optional.
 	MemoryOverhead *string `json:"memoryOverhead,omitempty"`
+	// GPU is the number of nvidia.com/gpu to request for the pod
+	// Optional.
+	GPU *int64 `json:"gpu,omitempty"`
```
Comment (liyinan926) [inline pkg/apis/sparkoperator.k8s.io/v1beta1/types.go]: We can make a struct type that looks like the following to support all GPU types from all vendors:
```go
type GPUSpec struct {
    Name string `json:"name"`
    Quantity int64 `json:"quantity"`
}
```

Then this field becomes:
```go
GPU *GPUSpec `json:"gpu,omitempty"`
```

File: pkg/webhook/patch_test.go [hunk 2]
```diff
@@ -676,6 +678,152 @@ func TestPatchSparkPod_Sidecars(t *testing.T) {
...
+			},
+			Executor: v1beta1.ExecutorSpec{
+				SparkPodSpec: v1beta1.SparkPodSpec{
+					GPU: &gpu,
+				},
+			},
+		},
+	}
+	tests := []testcase{
+		{nil,
```
Comment (liyinan926) [inline pkg/webhook/patch_test.go]: `nil` should be on a new line.

**Comment by tkanng** [comment]
Thank you for your patience! The latest commit adds `GPUSpec` type and refines corresponding unit tests. Users can specify `gpu` field, just like this:
```yaml
  executor:
   # cores: 1
    instances: 1
   # memory: "512m"
    gpu: 
      name: example.com/gpu 
      quantity: 1
```

File: pkg/webhook/patch.go [hunk 3]
```diff
@@ -18,6 +18,8 @@ package webhook
 
 import (
 	"fmt"
+	"k8s.io/apimachinery/pkg/api/resource"
```
Comment (liyinan926) [inline pkg/webhook/patch.go]: Nit: `k8s.io` imports should go immediately under built-in ones.

File: pkg/webhook/patch_test.go [hunk 4]
```diff
@@ -676,6 +678,199 @@ func TestPatchSparkPod_Sidecars(t *testing.T) {
 	assert.Equal(t, "sidecar2", modifiedExecutorPod.Spec.Containers[2].Name)
 }
 
+func TestPatchSparkPod_GPU(t *testing.T) {
+
```
Comment (liyinan926) [inline pkg/webhook/patch_test.go]: Nit: empty line can be removed.

File: pkg/apis/sparkoperator.k8s.io/v1beta1/types.go [hunk 5]
```diff
@@ -499,6 +502,13 @@ type PrometheusSpec struct {
 	Configuration *string `json:"configuration,omitempty"`
 }
 
+type GPUSpec struct {
+	// Name is GPU resource name, such as: nvidia.com/gpu or amd.com/gpu
+	Name string `json:"name"`
+	// Quantity is the number of GPU to request for driver or executor.
```
Comment (liyinan926) [inline pkg/apis/sparkoperator.k8s.io/v1beta1/types.go]: `number of GPUs`.

**Comment by tkanng** [comment]
Hi, the latest commit added user guide on how to use gpus and refined code format.     :)

**Comment by gyj0825** [comment]
How to add multiple GPU parameters to support bitfusion, like
limits:
   bitfusion.io/gpu-amount: 2
   bitfusion.io/gpu-percent: 50
thanks.

**Pull request merged**